Berry Farm Overhaul

HLXII opened the pull request:
See https://github.com/pokeclicker/pokeclicker/issues/902#issue-716952657

This is the first PR mostly for code review, as the last development hasn't been fully implemented yet (see above issue for current progress).

Functionality to check: 
- [ ] Mechanics
   - [ ] New Growth Implementation
   - [ ] Unlocking Plots
   - [ ] Mulch
   - [ ] Mutations
- [ ] UI
   - [ ] Farm UI
   - [ ] Berry/Mulch List
   - [ ] Berrydex Tab
   - [ ] Berrydex Modal
   - [ ] Berry Master UI
- [ ] NPCs
   - [ ] Kanto Berry Master Messages
   - [ ] Berry Master Trades
- [ ] Balance
   - [ ] Berry Growth Times
   - [ ] Berry Harvest Amounts
   - [ ] Berry Farm Tokens
   - [ ] Replant Rates
   - [ ] Mutation Rates
   - [ ] Trade Amounts

Aegyo reviewed (commented):
Partial review, I didn't get all the way through it tonight.

I'll take another look tomorrow

anapple96 commented on src/scripts/farming/Farming.ts:
```diff
@@ -5,63 +5,894 @@ class Farming implements Feature {
+        this.berryData[BerryType.Sitrus]    = new Berry(BerryType.Sitrus,   [150,300,450,600,1200],
+            8, 0.2, 30,
+            [0, 10, 10, 10, 10], BerryColor.Yellow,
+            ['Sitrus came from the same family as Oran. It is larget and smoother tasting than Oran.']);
```
Larget

Aegyo commented on src/scripts/farming/Farming.ts:
```diff
@@ -5,63 +5,894 @@ class Farming implements Feature {
+        // Coba
+        this.mutations.push(new FlavorMutation(.00001, BerryType.Coba,
+            [0, 10, 0, 15, 0], true,
+            'I\'ve heard that a berry will appear if its surroundings match its flavor profile! If I recall, it tasted a little dry and fairly bitter at the same time.',
```
```suggestion
            'I\'ve heard that a Berry will appear if its surroundings match its flavor profile! If I recall, it tasted a little dry and fairly bitter at the same time.',
```

Aegyo commented on src/scripts/farming/Farming.ts:
```diff
@@ -5,63 +5,894 @@ class Farming implements Feature {
+                change = true;
+            }
+            if (plot.notifications) {
+                plot.notifications.forEach(n => notifications.add(n));
```
```suggestion
                plot.notifications.forEach(notifications.add);
```

HLXII replied:
Tried this and didn't seem to work

Aegyo commented on src/scripts/farming/Farming.ts:
```diff
@@ -5,63 +5,894 @@ class Farming implements Feature {
+        if (notifications.size) {
+            notifications.forEach(function(n) {
+                this.handleNotification(n);
+            }, this);
```
I haven't tried it, but this can probably be changed to
```js
notifications.forEach(this.handleNotification)
```

HLXII replied:
same here

Aegyo commented on src/scripts/farming/Farming.ts:
```diff
@@ -5,63 +5,894 @@ class Farming implements Feature {
+        switch (type) {
+            case FarmNotificationType.Ripe:
+                Notifier.notify({
+                    message: 'A berry is ready to harvest!',
```
`berry` -> `Berry` on all of the notification messages

Aegyo commented on src/scripts/farming/Farming.ts:
```diff
@@ -70,86 +901,127 @@ class Farming implements Feature {
+
+        plot.die(true);
+
+        this.resetAuras();
```
Does this need to be wrapped in a `if (!suppressResetAura)`

Aegyo commented on src/scripts/farming/Plot.ts:
```diff
@@ -2,43 +2,434 @@ class Plot implements Saveable {
+                return '';
+            }
+            for (let i = 0;i < 5;i++) {
+                if (this.age < this.berryData.growthTime[i]) {
```
I think it would be clearer to do
```js
const growthTime = this.berryData.growthTime.find(t => this.age < t)
```
and then replace `this.berryData.growthTime[i]` with `growthTime` in the rest of the calculation, to avoid returning from the middle of a for loop.

Aegyo commented on src/scripts/farming/Plot.ts:
```diff
@@ -2,43 +2,434 @@ class Plot implements Saveable {
+
+            if (this.auraReplant() !== 1) {
+                if (auraStr) {
+                    auraStr += '<br/>';
```
Maybe just personal preference, but in cases like this, I like to collect all of the strings into an array and then
```js
return strings.join('<br/>')
```
at the end, instead of having to check if we already have stuff at each stage.

Aegyo commented on src/scripts/farming/Plot.ts:
```diff
@@ -2,43 +2,434 @@ class Plot implements Saveable {
+                return PlotStage.Seed;
+            }
+            for (let i = 0;i < 5;i++) {
+                if (this.age < this.berryData.growthTime[i]) {
```
Another
```js
return this.berryData.growthTime.find(t => this.age < t)

Aegyo commented on src/scripts/farming/Plot.ts:
```diff
@@ -2,43 +2,434 @@ class Plot implements Saveable {
+
+            if (this.stage() >= PlotStage.Taller && this.berryData.aura) {
+                if (tooltip) {
+                    tooltip += '<br/>';
```
Another collection of strings and join at the end, if you liked the suggestion.

Aegyo commented on src/scripts/farming/Plot.ts:
```diff
@@ -2,43 +2,434 @@ class Plot implements Saveable {
+        }
+        let newStage = PlotStage.Seed;
+        for (let i = 0;i < 5;i++) {
+            if (newAge <= growthStages[i]) {
```
A couple more loops that can be replaced with `.find`

Aegyo commented on src/scripts/farming/Plot.ts:
```diff
@@ -47,35 +438,63 @@ class Plot implements Saveable {
+        const plots = [];
+
+        const colIdx = index % 5;
+        const rowIdx = (index - colIdx) / 5;
```
These 5s could do with being constants somewhere, and also replaced wherever else is assuming that the farm is a 5x5

Aegyo commented on src/scripts/farming/aura/Aura.ts:
```diff
@@ -0,0 +1,68 @@
+    auraType: AuraType;
+    auraMultipliers: number[];
+
+    constructor(auraType: AuraType, auraMultipliers: number[]) {
```
You can write this line as
```js
constructor(
    public auraType: AuraType,
    public auraMultipliers: number[],
) { }
```
then you will not need the declarations above the constructor or the setting inside of the constructor, typescript will Just Know.

Aegyo commented on src/scripts/farming/aura/Aura.ts:
```diff
@@ -0,0 +1,68 @@
+        let label = '';
+        switch (this.auraType) {
+            case AuraType.Attract:
+                label = 'Attract: ';
```
This entire switch statement looks like it can be replaced with
```js
let label = `${AuraType[this.auraType]}: `;
```
Maybe a `GameHelper.humanifyString` if there might be a multiword AuraType added in the future.

Aegyo reviewed (commented):
Another partial review, only 21 more files to go...

Aegyo commented:
Also, thank you for showing me the word "drupelet" :+1:

RedSparr0w commented:
Haven't had a chance to look through any of this yet, but just checking that all the images have been minified with something like https://tinypng.com?

HLXII commented:
> Haven't had a chance to look through any of this yet, but just checking that all the images have been minified with something like https://tinypng.com?

Yep, though I didn't do that with the plant images since those are fairly small already. Not sure if it was worth it.

RedSparr0w commented:
Definitely worth it, as it doesn't change the image at all but reduces the file size by a decent amount.

For example:
\- | Before | After
:-- | :--: | :--:
Image | ![AllTreeSprout](https://user-images.githubusercontent.com/7288322/97496660-2767bf00-19ce-11eb-9714-7ea3248466c6.png) | ![AllTreeSprout (1)](https://user-images.githubusercontent.com/7288322/97496664-2898ec00-19ce-11eb-86c2-43aac4ab2d6d.png)
Size | 434 B | 343 B (-21%)
Image | ![ApicotTreeBerry](https://user-images.githubusercontent.com/7288322/97496766-53834000-19ce-11eb-9cc7-4854d50121b1.png) | ![ApicotTreeBerry (1)](https://user-images.githubusercontent.com/7288322/97496770-541bd680-19ce-11eb-8f68-090d48864947.png)
Size | 662 B | 494 B (-25%)

HLXII commented on src/scripts/farming/BerryDeal.ts:
```diff
@@ -0,0 +1,166 @@
+
+    public static getDeals(region: GameConstants.Region) {
+        switch (region) {
+            case GameConstants.Region.johto:
```
Removed this to instead set the available regions manually, and also split the berry deals into a recordd, so no more slicing

HLXII commented on src/scripts/farming/BerryDeal.ts:
```diff
@@ -0,0 +1,166 @@
+        return berryList[Math.floor(berryList.length * SeededRand.next())];
+    }
+
+    private static randomAmount(base: number, variance: number): number {
```
Found SeededRand has intBetween, which I can just use instead

RedSparr0w commented:
Awesome work, definitely makes the Farm a lot more interesting!

<hr>

What is the purpose of harvesting the berries?
It seems to always remove the berry, where as if you just leave it to "die", it will possibly auto replant and grow more berries.

<hr>

Maybe add a notification for when you unlock a new berry _(doesn't need a setting to disable/enable)_

<hr>

I think we could probably remove the `shovel all` button, as I feel like it will get miss clicked too often.
Maybe move the shovel button to the left near the berries, so it's more obvious that it changes the selection between the berries and the shovel.
_Which could also allow us to add an extra row of plots?_
![image](https://user-images.githubusercontent.com/7288322/98051674-3dc3be00-1e99-11eb-807d-ef8156b27a3e.png)

<hr>

Not sure what's causing this, but when I open the berry master shop, there is quite a lot of lag when opening the shop, and while the shop is open.

<hr>

I purchased the `Squirt Bottle` but it doesn't seem to do anything, or be anywhere?

<hr>

Possibly change this to be a table:
![image](https://user-images.githubusercontent.com/7288322/98051074-f557d080-1e97-11eb-9c0c-5e3a3b355f6e.png)

<hr>

I think we also need to either take these berries away from the player and replace them with tokens or something else,
Or just unlock them if they already have some:
![image](https://user-images.githubusercontent.com/7288322/98051259-597a9480-1e98-11eb-9b65-c01a35971189.png)

<hr>

Maybe align the text to the left:
![image](https://user-images.githubusercontent.com/7288322/98051492-e6255280-1e98-11eb-85aa-f43e71fc6dfb.png)
![image](https://user-images.githubusercontent.com/7288322/98051622-208eef80-1e99-11eb-8570-26fdb8db7130.png)

<hr>

I'm not sure what has happened here, maybe the berry images are different sizes and not square?
![image](https://user-images.githubusercontent.com/7288322/98051873-a448dc00-1e99-11eb-85e5-550426c48d44.png)

HLXII commented:
Harvesting Berries returns the full harvest amount. If you let the Berry plant die, you'll get half of the Berries, and there's a chance for the Berry to replant itself. This allows for passive Berry gain if you don't want to micromanage the Farm, and also rewards active play with full harvests.

Added a notification

Removed the Shovel All button. I've moved the Berry Shovel option under the Berry/Mulch element.
![image](https://user-images.githubusercontent.com/33099029/98057570-30122800-1e08-11eb-903d-20eb31103a85.png)
I didn't want to move it into the Berry List since that would require additional work for handling change Berry pages.

I'm not sure about the Berry Modal lag either. I removed the footer as well as the Berry Trade tab and it seems to still happen. I assume it might be some Knockout problem but idk.

The SquirtBottle is an Oak Item that's unlocked after purchasing from the store. I might add a description hover tooltip for it.

Made the rest of the UI changes.

RedSparr0w commented:
> I'm not sure about the Berry Modal lag either. I removed the footer as well as the Berry Trade tab and it seems to still happen. I assume it might be some Knockout problem but idk.

It only seems to happen while in the Shop tab.
Did you use the same layout and everything as the other shops?

> I've updated Update.ts to unlock the already obtained berries, however I'd need to update the dev version for that code to run.

Looks like it's working now, just wasn't running the update function before as the `package.json` version wasn't `0.5.9`

RedSparr0w commented:
After playing through the Farm from a fresh save,
I think we probably should just reset farm progress.
Otherwise you are just missing/skipping a ton of the content for Farm.

Possibly:
- Refund Farm tokens spent on plots
- Convert all their berries to Farm tokens at a lowish rate _(10 - 100 berries per token)?_

RedSparr0w commented on src/scripts/farming/BerryType.ts:
```diff
@@ -1,11 +1,85 @@
+    Bluk,
+    Nanab,
+    Wepear,
+    Pinap,
+
+    Figy,
+    Wiki,
+    Mago,
+    Aguav,
+    Iapapa,
+
+    Lum,
```
Is there any reason for all the extra spacings in this file?
```suggestion
    Pinap,
    Figy,
    Wiki,
    Mago,
    Aguav,
    Iapapa,
    Lum,
```

HLXII replied:
I separated them into groups for determining properties. The spaces can be removed.

RedSparr0w commented on src/scripts/oakItems/OakItems.ts:
```diff
@@ -18,14 +23,26 @@ class OakItems implements Feature {
+                false, [1.50, 1.60, 1.70, 1.80, 1.90, 2.00], 1, 80, 10),
+            new OakItem(OakItems.OakItem.Cell_Battery, 'Cell Battery', 'More passive mining energy regen',
+                false, [1.5, 1.6, 1.7, 1.8, 1.9, 2], 1, 90, 20),
+            new BoughtOakItem(OakItems.OakItem.Squirtbottle, 'Squirtbottle', 'Increases the chance of berry mutations', 'Johto Berry Master',
+                true, [1.15, 1.3, 1.45, 1.6, 1.75], 1, 5, undefined, undefined, AmountFactory.createArray([500, 1000, 2000, 3000, 5000], GameConstants.Currency.farmPoint)),
+            new BoughtOakItem(OakItems.OakItem.Sprinklotad, 'Sprinklotad', 'Increases the chance of berry replants', 'Hoenn Berry Master',
+                true, [1.5, 1.75, 2, 2.25, 2.5], 1, 7, undefined, undefined, AmountFactory.createArray([500, 1000, 2000, 3000, 5000], GameConstants.Currency.farmPoint)),
```
Are these 2 the wrong way around?

As with the Sprinklotad, if the multiplier is 2+ then it's a 100% replant rate with all the Kanto berries.
Also seems to be able upgrade way too fast IMO.
```suggestion
            new BoughtOakItem(OakItems.OakItem.Squirtbottle, 'Squirtbottle', 'Increases the chance of berry mutations', 'Johto Berry Master',
                true, [1.15, 1.3, 1.45, 1.6, 1.75], 1, 5, undefined, undefined, AmountFactory.createArray([500, 1000, 2000, 3000, 5000], GameConstants.Currency.farmPoint)),
            new BoughtOakItem(OakItems.OakItem.Sprinklotad, 'Sprinklotad', 'Increases the chance of berry replants', 'Hoenn Berry Master',
                true, [1.5, 1.75, 2, 2.25, 2.5], 1, 7, undefined, undefined, AmountFactory.createArray([500, 1000, 2000, 3000, 5000], GameConstants.Currency.farmPoint)),
```

HLXII replied:
The Oak Items were implemented before I fully set all the properties for the berries. I'm always up for rebalancing (mostly because I didn't do that many test runs to check if they were balanced to being with). 

Experience required probably needs to go way up, especially since there are a lot more berries to work with now, so we can extend the Berry mechanic way into the next regions (aka slower growth).

We do want to balance the Sprinklotad to ensure that at the end game the player can have 100 replant so the farm can become fully passive. Whether that means 2x, or something less with Replant aura berries around is to be decided still.

RedSparr0w replied:
IMO it shouldn't ever be a 100% replant chance from just the Oak Item, especially with how quickly the oak items level up.

RedSparr0w replied:
We could probably also increase the upgrade cost, as that also seems quite low.

RedSparr0w replied:
Maybe these amount would work okay, What do you think?
```suggestion
            new BoughtOakItem(OakItems.OakItem.Squirtbottle, 'Squirtbottle', 'Increases the chance of berry mutations', 'Johto Berry Master',
                true, [1.5, 1.75, 2, 2.25, 2.5], 1, 10, undefined, undefined, AmountFactory.createArray([1000, 2000, 5000, 10000, 20000], GameConstants.Currency.farmPoint)),
            new BoughtOakItem(OakItems.OakItem.Sprinklotad, 'Sprinklotad', 'Increases the chance of berry replants', 'Hoenn Berry Master',
                true, [1.15, 1.3, 1.45, 1.6, 1.75], 1, 2, undefined, undefined, AmountFactory.createArray([1000, 2000, 5000, 10000, 20000], GameConstants.Currency.farmPoint)),
```

RedSparr0w commented on src/scripts/items/EvolutionStone.ts:
```diff
@@ -59,3 +59,35 @@ ItemList['Razor_fang']    = new EvolutionStone(GameConstants.StoneType.Razor_fan
 ItemList['Sachet']         = new EvolutionStone(GameConstants.StoneType.Sachet, 2500, undefined , 'Sachet');
 ItemList['Whipped_dream']  = new EvolutionStone(GameConstants.StoneType.Whipped_dream, 2500, undefined , 'Whipped Dream');
+
+namespace EvoItems {
+    export enum EvoItem {
+        'Fire_stone' = 0,
+        'Water_stone',
+        'Thunder_stone',
+        'Leaf_stone',
+        'Moon_stone',
+        'Sun_stone',
+        'Trade_stone',
+        'Dragon_scale',
+        'Metal_coat',
+        'Kings_rock',
+        'Upgrade',
+        'Soothe_bell',
+        'Deepsea_tooth',
+        'Deepsea_scale',
+        'Dawn_stone',
+        'Dusk_stone',
+        'Shiny_stone',
+        'Dubious_disc',
+        'Electirizer',
+        'Magmarizer',
+        'Protector',
+        'Reaper_cloth',
+        'Razor_claw',
+        'Razor_fang',
+        'Prism_scale',
+        'Sachet',
+        'Whipped_dream',
+    }
+}
```
I don't think we should do this, as it just creates more places that need to be edited when we add a new evo item, and it's the same output as `GameConstants.StoneType` (without the `None`).

RedSparr0w commented on src/scripts/items/BattleItem.ts:
```diff
@@ -20,4 +20,15 @@ ItemList['xClick']          = new BattleItem(GameConstants.BattleItemType.xClick
\ No newline at end of file
+ItemList['Lucky_incense']   = new BattleItem(GameConstants.BattleItemType.Lucky_incense, '+50% Bonus to money gained', 2000, undefined, 'Lucky Incense');
+
+namespace BattleItems {
+    export enum BattleItem {
+        'xAttack' = 0,
+        'xClick',
+        'Lucky_egg',
+        'Token_collector',
+        'Item_magnet',
+        'Lucky_incense',
+    }
+}
```
I don't think we should do this, as it just creates more places that need to be edited when we add a new battle item.
If we need a numeric enum, we should probably modify `GameConstants.BattleItemType` to fit our needs.

HLXII replied:
Fixed, though I couldn't use SeededRand.fromEnum, since the enum values for BattleItemType aren't numbers.

RedSparr0w commented on src/scripts/items/ShovelItem.ts:
```diff
@@ -0,0 +1,18 @@
+    }
+}
+
+// TODO: HLXII - Balance cost based on highest yield strategy
+ItemList['Berry_Shovel']   = new ShovelItem(500, 'Berry Shovel');
```
Would be good to get all the TODOs done before merge as well where possible,
Not sure what this should cost though.
```suggestion
ItemList['Berry_Shovel']   = new ShovelItem(500, 'Berry Shovel');
```

RedSparr0w reviewed (approved):
Changes look great :tada:

Thanks for the PR!
